tests/Microsoft.Data.DataFrame.Tests/DataFrameTests.cs

```diff
@@ -744,5 +772,156 @@ public void TestJoin()
+                    Assert.Equal(originalColumn[r].ToString(), headColumn[verify[r]].ToString());
+                }
+            }
+            for (int c = 0; c < count.ColumnCount; c++)
+            {
+                BaseColumn originalColumn = df.Column(c);
+                if (originalColumn.Name == "Bool")
+                    continue;
+                BaseColumn headColumn = head[originalColumn.Name];
+                Assert.Equal(originalColumn[5], headColumn[verify[5]]);
+            }
+            Assert.Equal(6, head.RowCount);
+
+            DataFrame tail = df.GroupBy("Bool").Tail(3);
+            Assert.Equal(6, tail.RowCount);
+            List<int> originalColumnVerify = new List<int>() { 6, 8, 7, 9 };
+            List<int> tailColumnVerity = new List<int>() { 1, 2, 4, 5 };
+            for (int r = 0; r < 4; r++)
+            {
+                for (int c = 0; c < count.ColumnCount; c++)
+                {
+                    BaseColumn originalColumn = df.Column(c);
+                    BaseColumn tailColumn = tail[originalColumn.Name];
+                    Assert.Equal(originalColumn[originalColumnVerify[r]].ToString(), tailColumn[tailColumnVerity[r]].ToString());
+                }
+            }
+
+            DataFrame max = df.GroupBy("Bool").Max();
+            Assert.Equal(2, max.RowCount);
+            for (int r = 0; r < 2; r++)
+            {
+                for (int c = 0; c < count.ColumnCount; c++)
+                {
+                    BaseColumn originalColumn = df.Column(c);
+                    if (originalColumn.Name == "Bool" || originalColumn.Name == "Char")
+                        continue;
+                    BaseColumn maxColumn = max[originalColumn.Name];
+                    Assert.Equal(((long)(r == 0 ? 8 : 9)).ToString(), maxColumn[r].ToString());
+                }
+            }
+
+            DataFrame min = df.GroupBy("Bool").Min();
+            Assert.Equal(2, min.RowCount);
+
+            DataFrame product = df.GroupBy("Bool").Product();
+            Assert.Equal(2, product.RowCount);
+
+            DataFrame sum = df.GroupBy("Bool").Sum();
+            Assert.Equal(2, sum.RowCount);
+            for (int r = 0; r < 2; r++)
+            {
+                for (int c = 0; c < count.ColumnCount; c++)
+                {
+                    BaseColumn originalColumn = df.Column(c);
+                    if (originalColumn.Name == "Bool" || originalColumn.Name == "Char")
+                        continue;
+                    BaseColumn minColumn = min[originalColumn.Name];
+                    Assert.Equal("0", minColumn[r].ToString());
+
+                    BaseColumn productColumn = product[originalColumn.Name];
+                    Assert.Equal("0", productColumn[r].ToString());
+
+                    BaseColumn sumColumn = sum[originalColumn.Name];
+                    Assert.Equal("20", sumColumn[r].ToString());
+                }
+            }
+        }
+
+        [Fact]
```

Comment:
The template types are the ones changing here. Is there a way to use Theory and InlineData for this?

Reply:
One option would be to use `[Theory]` and `[MemberData]`. Something like:

```

public static IEnumerable<object[]> GroupByDataFrames()
{
    MakeDataFrame<byte, bool>(10, false), 
    MakeDataFrame<char, bool>(10, false),
    ....
}

[Theory]
[MemberData(nameof(GroupByDataFrames))]
public void TestGoupByDifferentColumnTypes(DataFrame df)
{
    DataFrame grouped = df.GroupBy("Column1").Count();
    Assert.Equal(2, grouped.RowCount);
}

```

I'm not sure if it is significantly better than what you have here, but that would be a way to do it with a Theory.

---
In reply to: [304140369](https://github.com/dotnet/corefxlab/pull/2696#discussion_r304140369) [](ancestors = 304140369)